Merge back 'chore_release-pd-8.5.0' into 'edge'

ddcc4 opened the pull request:
This brings the hotfixes from `chore_release-8.5.1` into `edge`.

There were merge conflicts in `protocol-designer/fixtures/protocol/8/example_1_1_0MigratedToV8.json` from @ncdiehl's PR #19041, which @rclarke0 had also changed in PR #18725. I went with @ncdiehl's version since his was more recent. Let's see if that works ...

sanni-t committed:
docs(api): liquid classes docstrings and descriptions updates (#18996)

# Overview

Fixes some docstrings in liquid-class related functions and some
descriptions of the liquid class schema fields.

## Changelog

- Docstring updates for:
- `transfer_with_liquid_class`, `consolidate_with_liquid_class`,
`distribute_with_liquid_class`
  - `get_liquid_class`, `define_liquid_class`
- Liquid classes json schema field descriptions

## Risk assessment

None. Docs fixes only

---------

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

ecormany committed:
chore: additional release note for air gap volume bug fix (#19017)

# Overview

Adding a release note entry for the fix to
[AUTH-2138](https://opentrons.atlassian.net/browse/AUTH-2138).

## Test Plan and Hands on Testing

See it in the next alpha.

## Changelog

+1

## Review requests

Good, accurate description of the fix?

## Risk assessment

nil

[AUTH-2138]:
https://opentrons.atlassian.net/browse/AUTH-2138?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ecormany committed:
chore: release note for floating point distribute bug fix (#19033)

# Overview

Add a line for
[AUTH-2151](https://opentrons.atlassian.net/browse/AUTH-2151) in the
8.5.1 release notes.

## Test Plan and Hands on Testing

n/a

## Changelog

- 1 new bullet
- Simplified note for air gap bug per [post-merge discussion on
#19017](https://github.com/Opentrons/opentrons/pull/19017#discussion_r2228692853)

## Review requests

good words?

## Risk assessment

none

[AUTH-2151]:
https://opentrons.atlassian.net/browse/AUTH-2151?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ncdiehl11 committed:
fix(protocol-designer): migrate flow rates that exceed pipette capacity (#19041)

Old protocols specified an inaccurately permissive flow rate range for
aspirate, dispense, and blowout, which could exceed the pipette's
physical capacity for plunger movement. We have since mirrored the
hardware-controller's logic for determining the max flow rate for the
pipette in question, based on target volume, tip type, correction
volume, flow rate type, etc. However, that logic was wired up for
creating new forms or setting liquid class values, not for migrating
older moveLiquid/mix steps that may have specified an invalid flow rate.
This PR adds the logic for "clipping" the flow rate to the pipette's
physical maximum during migration.

Closes RQA-4450, RQA-4451

sanni-t committed:
fix(api): post-dispense air gaps in LC based transfers (#19025)

Closes AUTH-2138

# Overview

We were previously using `airGapByVolume` of post-aspirate retract
property for finding air gaps to use during post-dispense retracts. This
PR fixes that so we use the `airGapByVolume` property of a
post-single-dispense or a post-multi-dispense in such cases.

Additionally, when adding air gaps in-between multi-dispenses (when
conditioning volume is zero), we should be using the total tip volume as
the lookup volume for `airGapByVolume` but we were just using `0`uL.

## Risk assessment

Medium. This fixes the air gaps in a lot of cases, but the impact of
this change is not too high practically speaking.

ncdiehl11 committed:
fix max disposal volume logic (#19045)

Correctly get tiprack def URI passed to
`getMaxDisposalVolumeForMultiDispense` in `DisposalField`

Closes [RQA-4442](https://opentrons.atlassian.net/browse/RQA-4442)

ddcc4 committed:
fix(api): do not rely on inexact volume in tip to determine last dispense (#19030)

# Overview

In `distribute_with_liquid_class()`, after we dispense to the last well
of a multi-dispense, we're supposed to blow out the disposal volume.
This is triggered by calling
`retract_during_multi_dispensing(is_last_retract=True)`.

However, we were computing `is_last_retract` as:
```
components_executor.retract_during_multi_dispensing(
    is_last_retract=tip_starting_volume - volume == disposal_volume,
)
```
where `tip_starting_volume` is the current volume in the tip before the
dispense, `volume` is the amount we're about to dispense, and
`disposal_volume` is the disposal volume.

This equality becomes **inexact** when `tip_starting_volume`, `volume`,
and `disposal_volume` are not nice round numbers, which often happens
with liquid classes because the `disposal_volume` is interpolated.

This floating point bug causes `is_last_retract` to be `False`, which
causes us to never dispose of the disposal volume. (AUTH-2151)

This fix changes the code to no longer rely on `tip_starting_volume` to
determine whether we are about to do the last dispense. Instead, in the
loop that performs the multi-dispense, we simply see if we're in the
last iteration of the loop.

## Test Plan and Hands on Testing

I updated the unit tests. Note that we already have existing unit tests
for both the `is_last_retract=False` and `is_last_retract=True` cases.

After my change, I was able to delete
`decoy.when(get_aspirated_volume(...)).then_return(...)` from the unit
tests, which ensures that our implementation no longer looks at the
current volume in the tip.

I also manually analyzed this protocol to check that we are blowing out
the disposal volume:
```
from opentrons import protocol_api, types

requirements = {"robotType": "Flex", "apiLevel": "2.24"}

def run(protocol: protocol_api.ProtocolContext) -> None:
    tip_rack = protocol.load_labware(
        "opentrons_flex_96_tiprack_200ul",
        location="C2",
        namespace="opentrons",
    )
    well_plate = protocol.load_labware(
        "opentrons_96_wellplate_200ul_pcr_full_skirt",
        location="B1",
        namespace="opentrons",
    )
    pipette = protocol.load_instrument(
        "flex_1channel_1000", "left", tip_racks=[tip_rack],
    )
    waste_chute = protocol.load_waste_chute()

    liquid_class = protocol.get_liquid_class(name="water")
    transfer_props = liquid_class.get_for(pipette, tip_rack)
    transfer_props.multi_dispense.retract.blowout.location = "trash"
    # Before this fix, we would not blow out this disposal volume:
    transfer_props.multi_dispense.disposal_by_volume.set_for_all_volumes(1.1)

    pipette.distribute_with_liquid_class(
        liquid_class=liquid_class,
        volume=10,
        source=[well_plate["A1"]],
        dest=[well_plate["G11"], well_plate["H12"]],
        new_tip="once",
        return_tip=True,
    )
```

## Risk assessment

Medium. This is rather complicated code.

ddcc4 committed:
test(api): fix test_order_of_water_distribution_steps_using_multi_dispense_without_conditioning_volume() (#19049)

# Overview

In PR #19030, I introduced the new argument `is_last_dispense_in_tip` to
`dispense_liquid_class_during_multi_dispense()` at the same time
@sanni-t in PR #19025 added the new test case
`test_order_of_water_distribution_steps_using_multi_dispense_without_conditioning_volume()`.

When both changes were checked in together, the new test case did not
have the new argument, causing the test to fail.

This fix updates the test case to expect the new `is_last_dispense_in_tip`
argument.

## Test Plan and Hands on Testing

Ran the test locally, and run it in CI.

## Risk assessment

Low. Just fixes a test, has no effect on behavior.

ddcc4 committed:
Merge back 'chore_release-8.5.1' into 'chore_release-pd-8.5.0' (#19037)

ncdiehl11 committed:
fix(protocol-designer, step-generation): fix `getTransferPlanAndReferenceVolumes` multiAspirate logic (#19047)

This PR fixes logic to correctly determine source wells to fit in tip
for a multiAspirate transfer in `getTransferPlanAndReferenceVolumes`
planning utility. We pass the number of source wells to find the correct
volume in the tip, considering the max working volume of the tip.

Closes RQA-4431

ncdiehl11 committed:
fix(protocol-designer): populate blowout flow rate for no liquid class (#19048)

For both OT-2 and Flex moveLiquid forms with no liquid class selected,
we should populate the default blowout flow rate, even though the
blowout/disposal checkbox begins unchecked. For OT-2, we use the
pipette's liquid specs default blowout flow rate. For Flex, we use the
reference liquid class's (water) blowout flow rate if it exists. We
fallback to the pipette's liquid specs default flow rate for Flex.

Closes RQA-4453

ncdiehl11 committed:
fix(protocol-designer): populate blowout flow rate, remove disposal volume warning (#19046)

cherry-picked commits accidentally targeted edge in #19044

This PR ensures blowout flow rate is populated for "don't use a liquid
class" transfers. It also removes a legacy warning for disposal volume
below recommended value (liquid classes specify values below the 5ul
minimum frequently).

Closes RQA-4425

sanni-t reviewed (approved):
Backend changes look good

ncdiehl11 committed:
fix(protocol-designer): fix duplicate property after merge (#19056)

A sloppy merge resulted in a duplicate object property in a form change
handler util. This PR removes that duplicate.

ddcc4 committed:
Merge back 'chore_release-pd-8.5.0' into 'edge' (#19052)

This brings the hotfixes from `chore_release-8.5.1` into `edge`.

Manually edited merge conflicts in: `protocol-designer/fixtures/protocol/8/example_1_1_0MigratedToV8.json`, `protocol-designer/src/steplist/formLevel/handleFormChange/utils.ts`, `app/src/organisms/ODD/QuickTransferFlow/QuickTransferAdvancedSettings/BlowOut.tsx`.

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal merge-pd-8.5.0-into-edge snapshots (#19053)

This PR was requested on the PR https://github.com/Opentrons/opentrons/pull/19052

Co-authored-by: ddcc4 <46395074+ddcc4@users.noreply.github.com>
